pkg/operator/api/components.go

```diff
@@ -0,0 +1,240 @@
+/*
+Copyright 2024 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package api
+
+import (
+	"context"
+	b64 "encoding/base64"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/google/uuid"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	commonapi "github.com/dapr/dapr/pkg/apis/common"
+	componentsapi "github.com/dapr/dapr/pkg/apis/components/v1alpha1"
+	httpendpointsapi "github.com/dapr/dapr/pkg/apis/httpEndpoint/v1alpha1"
+	operatorv1pb "github.com/dapr/dapr/pkg/proto/operator/v1"
+)
+
+type ComponentUpdateEvent struct {
+	Component *componentsapi.Component
+	EventType operatorv1pb.ResourceEventType
+}
+
+func (a *apiServer) OnComponentUpdated(ctx context.Context, eventType operatorv1pb.ResourceEventType, component *componentsapi.Component) {
+	a.connLock.Lock()
+	var wg sync.WaitGroup
+	wg.Add(len(a.allConnUpdateChan))
+	for _, connUpdateChan := range a.allConnUpdateChan {
+		go func(connUpdateChan chan *ComponentUpdateEvent) {
+			defer wg.Done()
+			select {
+			case connUpdateChan <- &ComponentUpdateEvent{
+				Component: component,
+				EventType: eventType,
+			}:
+			case <-ctx.Done():
+			}
+		}(connUpdateChan)
+	}
+	wg.Wait()
+	a.connLock.Unlock()
+}
+
+// ComponentUpdate updates Dapr sidecars whenever a component in the cluster is modified.
+// TODO: @joshvanl: Authorize pod name and namespace matches the SPIFFE ID of
```

Comment:
Will this require a K8s API server call?

Reply:
No. This comment isn't written very well as pod name isn't involved- it is app ID we care about. It is referring to server-side Component scope filtering, implemented [here](https://github.com/dapr/dapr/pull/7615).

Component updates are broadcast to the correct streaming clients based on their identity derived via mTLS, stored in-memory.